### DISCUSSION TITLE
React Query vs supabase subscription

### DISCUSSION THREAD

**Discussion by geniusit** [supabase/supabase#5048] [Questions] [answered]
Hello all,

First of all a big THANK YOU for all your work. I like supabase very much.

Today I wonder about the utility of React Query vs supabase subscription.

The use case I describe here is to get a list of countries and keep it as fresh as possible. 

So with subscription you have this code : 

```typescript

CountryList.tsx

import { useEffect, useState } from "react";
import { useCountry } from "../hooks/useCountry";
import { supabase } from "../utils/supabaseClient";

const CountryList = () => {

    const countries = useCountry();

  return (
    <>
      {Array.from(countries.values()).map((c) => (
        <div key={c.id}>{c.name}</div>
      ))}
    </>
  );
};

export default CountryList;

---------------------------------------------------------------------------------------

useCountry.tsx

import { useState, useEffect } from "react"
import {Country} from "../types/country.type"
import { supabase } from "../utils/supabaseClient";

export const useCountry= () => {

    const [countries, setCountries] = useState(new Map<number, Country>())

    const [newOrUpdatedCountry, handleNewOrUpdatedCountry] = useState<Country | null>(null);

    useEffect(() => {
        const mySubscription = supabase
          .from<Country>("countries")
          .on("*", (payload) => {
            handleNewOrUpdatedCountry(payload.new);
          })
          .subscribe();
      }, []);
    
      useEffect(() => {
        if (newOrUpdatedCountry) {
          setCountries((prev) =>
            new Map(prev).set(newOrUpdatedCountry.id, newOrUpdatedCountry)
          );
        }
      }, [newOrUpdatedCountry]);

      useEffect(() => {
        getCountries();
        
      }, []);

      async function getCountries() {
        try {
          let { data, error, status } = await supabase.from<Country>("countries");
    
          if (error && status !== 406) {
            throw error;
          }
    
          if (data) {
            data.forEach((d: Country) => {
              setCountries((prev) => new Map(prev).set(d.id, d));
            });
          }
        } catch (error: any) {
          alert(error.message);
        }
      }

    return countries
}
```

while with React Query you have only this : 

```typescript

CountryList.tsx

import useCountry from "../hooks/useCountry";
import {Country} from "../types/country.type"

const CountryList = () => {

    const {isLoading, data, isError, error, isFetching} = useCountry();

    if(isLoading){
        return <h2>Loading ...</h2>
    }

    if(isError){
        return <h2>{error?.message}</h2>
    }

  return (
    <>
        <h2>Countries list</h2>
        {
            data?.map((country:Country) => {
                return <div key={country.id}>{country.name}</div>
            })
        }
      
    </>
  );
};

export default CountryList;

--------------------------------------------------------------------

useCountry.tsx

import { useQuery } from "react-query";
import { Country } from "../types/country.type";
import { supabase } from "../utils/supabaseClient";

const getCountries = async () => {
  let { data, error, status } = await supabase.from<Country>("countries");

  if (error && status !== 406) {
    throw new Error(error.message);
  }

  if (!data) {
    throw new Error("No country found");
  }

  return data;
};

export default function useCountry() {
  return useQuery<Country[], Error>("countries", () => getCountries());
}


```

I think that I must choose between React Query and supabase  subscription.
As you know, with supabase subscription, the database update statement immediately reflect on the react application and I like that very much!

But I don't really know what to do ... React query is also really nice!
Does it make sense to mix both react query and supabase subscription ?

Please could you tell me what do you think about that ?

Thank you!!

**Comment by GaryAustin1** [comment, marked as answer]
I was waiting on confirmation from Supabase on something before responding.  

If you want to use realtime subscriptions, and expect to receive all updates to a table, you need to also handle subscription errors and reload the table and restart subscription on reconnects.  

Subscriptions can error based on loss of internet connection even for brief instances and will try and reconnect.  If they reconnect and you are not monitoring the subscription for errors, you will potentially miss any updates to the table during the break in connection and not know it.  
Connection errors can occur easily on mobile devices, but they also currently occur on at least Chrome and latest Edge browsers when your tab is not top of screen for more than 10 minutes.  
So if you need reliable realtime updates you must capture the subscription status and take action to reload data.

I do something like:
```
            .subscribe((status)=>{
                console.log('subscribe = status',status);
                if (status==='SUBSCRIBED') {
                    online.set(true);
                }
                else if (status !== 'CLOSED') {
                    online.set(false)
                }
            })
```
Then take action on online state change.  I also just close out subscriptions when the browser visibilityState is 'hidden' for more than 5 minutes and restart everything when 'visible'.

Another option is to just reload tables every once in awhile to make sure they are current, but that is sort of what React Query does if I remember right.

**Reply by cmnord** [reply]
> take action on online state change

What action do you take on that state change specifically? Do you call `.subscribe()` again, or something else? I'm writing a `useSubscribe` React hook right now and trying to figure out how to re-connect after, for example, a browser client leaves the tab for a couple minutes.

My implementation so far: https://github.com/cmnord/jep/blob/main/app/engine/use-game-engine.ts#L96-L170

**Comment by captainR0bbo** [comment]
Thank you for this... also looking at if I should even be using react-query with supabase if I want users to receive realtime updates.

**Comment by bcmendis** [comment]
Hi, I have a similar dilemma.
What did you guys end up going with?
Did you use both tanstack query and supabase subscriptions?

**Reply by geniusit** [reply]
I ve choose to use react-query. In general, I try to evict to use some black box tool like supabase. Realtime subscription are good but also may become headach when some network failure happen for example. How to recover? I also like the swr vercel library which is, like react-query a wrapper arround axios or the fetch api. I like to write my back-end REST services with Java framework like springboot or quarkus. I like to use hibernate and jpa or reactive postgresql to query my database. To me thats the perfect combo. Sorry for supabase even if im sure it s a super tool.

**Reply by bcmendis** [reply]
Understandable, though I was looking into using both.

My app involves multiple users adding information to the database, and each user needs to see fresh data asap, to avoid the risk of adding duplicate/incorrect information or the app returning errors, which would result in bad UX.

I'm gonna see if I can feasibly incorporate both tanstack query and supabase subscriptions.

Thanks anyways :)

**Reply by AfricanBongo** [reply]
@bcmendis I know I am probably late to offer my two cents

I figured out how to use both tools in my own project where I am trying to monitor file sharing between multiple users

/hooks/useFiles.tsx:

```typescript jsx
export default function useFiles({
  ownerId,
  reloadKey,
  onUpdateCallback,
}: Props): UseQueryResult<DicomFile[], Error> {
  useEffect(() => {
    if (onUpdateCallback) {
      listenToRealtimeUpdates(onUpdateCallback);
    }
  }, []);

  return useQuery({
    queryKey: ["files", reloadKey],
    queryFn: async () => {
      return getFiles(ownerId as string);
    },
    enabled: !!ownerId,
  });
}
```

That ```listenToRealtimeUpdates``` has the standard Supabase channel subscription and the ```onUploadCallback``` actually changes the query key each time an update event is triggered from Supabase. Therefore whenever another user changes the files or shares files with a user the changes are reflected in real-time.

Do note I am a fairly experienced developer but a React newbie so there might be better ways to do this.

Hope it helps 😄
